Examples/GraphColoring.lean

```diff
@@ -0,0 +1,232 @@
+    for_all U fun v =>
+      vertexColorClause v)
+  |> mapProp (by ext τ; apply forall_congr'; simp)
+
+def adjacentVertexesClauses (G : Graph n) [DecidableRel G.Adj] : VCnf n (edgeConstraints G) :=
+  ( let U := (Array.finRange n)
+    for_all U fun u =>
+    for_all U fun v =>
+    VEncCNF.guard (u ≠ v) fun _ =>
+    VEncCNF.guard (G.Adj u v) fun _ =>
+      seq[
+        addClause #[mkNeg <| blue   u, mkNeg <| blue   v],
+        addClause #[mkNeg <| red    u, mkNeg <| red    v],
+        addClause #[mkNeg <| green  u, mkNeg <| green  v],
+        addClause #[mkNeg <| yellow u, mkNeg <| yellow v]
+      ])
+  |> mapProp (by
+    ext τ
+    simp [edgeConstraints, adjacentVertexesGetDifferentColors, Clause.toPropFun]
+    apply forall_congr'; intro u
+    apply forall_congr'; intro v
+    aesop)
+
+def graphColoringCNF (G : Graph n) [DecidableRel G.Adj] : VCnf n (graphColoring G) :=
+  (seq[
+    vertexColorClauses n,
+    adjacentVertexesClauses G
+  ]).mapProp (by simp [graphColoring]; rfl)
+
+------------------------------------------------------------------------
+
+-- Now we prove that if there is a valid coloring, then we can produce an
+-- assignment on the variables for the graph
+
+def coloringAssignment (C : Coloring n 4) : PropAssignment (ColorVars n) :=
+  fun
+  | .blue   v => C v = ⟨0, by decide⟩
+  | .red    v => C v = ⟨1, by decide⟩
+  | .green  v => C v = ⟨2, by decide⟩
+  | .yellow v => C v = ⟨3, by decide⟩
+
+example (f : ℕ → Prop) (p : Fin 3) (h0 : f 0) (h1 : f 1) (h2 : f 2) : f p.val := by
+  fin_cases p; simp
+  all_goals assumption
+
+theorem splitFin4 (n : Fin 4) :
+  n = ⟨0, by decide⟩ ∨ n = ⟨1, by decide⟩ ∨
+  n = ⟨2, by decide⟩ ∨ n = ⟨3, by decide⟩ := by
+  fin_cases n <;> simp
+
+theorem coloringAssignment_exists_of_validColoring {G : Graph n} :
+  (∃ (C : Coloring n 4), isValidColoring G C) →
+  (∃ (τ : PropAssignment (ColorVars n)), τ |> graphColoring G) := by
+  rintro ⟨C, hC⟩
+  use coloringAssignment C
+  simp [graphColoring]
+  constructor
+  · simp [vertexColorConstraints, eachVertexGetsAColor, coloringAssignment]
+    intro v
+    rcases splitFin4 (C v) with (h | h | h | h)
+    <;> simp [h]
+  · simp [edgeConstraints, adjacentVertexesGetDifferentColors, coloringAssignment]
+    intro u v huv
+    have h_ne := hC huv
+    rcases splitFin4 (C u) with (hu | hu | hu | hu)
+    <;> simp [hu] at h_ne ⊢
+    <;> rcases splitFin4 (C v) with (hv | hv | hv | hv)
+    <;> simp [hv, ← not_and_or] at h_ne ⊢
+    <;> aesop
+
+-- Now generate an actual CNF for a graph object
+def K8 : Graph 8 := completeGraph (Fin 8)
+
+instance : DecidableRel K8.Adj := by
+  unfold DecidableRel
+  intro a b
+  by_cases h : Ne a b
+  · exact isTrue h
+  · exact isFalse h
+
+def K8_CNF : VCnf 8 (graphColoring K8) :=
+  graphColoringCNF K8
+
+def PrintTheCNF : IO Unit :=
+  let cnf := K8_CNF.val.toICnf
+  Solver.Dimacs.printFormula IO.print cnf
+
+-- #eval PrintTheCNF
+
+-- We trust the SAT solver and axiomatize the UNSAT result
```

Comment:
let's leave a TODO here for when we have a macro that handles introducing this axiom

Reply:
New comment left.